### PR TITLE
Feat 1872 - public contract collection & feat-1653 debug/update import

### PR DESCRIPTION
**Pull request by phpgeekfr**
functions set to keep up-to-date publicContract documents 
Update on movie spreadsheet for import

### CODE REVIEW

File: apps/backend-functions/src/contract.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+ */
+async function transformContractToPublic(contract: ContractDocument, versionToSkip?: string): Promise<void> {
+  return db.runTransaction(async tx => {
+    const publicContractSnap = await tx.get(db.doc(`publicContracts/${contract.id}`));
+
+    const versionSnap = await tx.get(db.doc(`contracts/${contract.id}/versions/_meta`));
+    const versionDoc = versionSnap.data();
+    let lastVersionDoc;
+    if (!!versionDoc) {
+      const versionToFetch = !versionToSkip || parseInt(versionToSkip, 10) !== parseInt(versionDoc.count, 10) ? versionDoc.count : parseInt(versionToSkip, 10) - 1;
```
Comment (ioaNikas) [inline apps/backend-functions/src/contract.ts]: Can you be more explicit or comment what you are doing here ? It's hard to understand from a noob perspective :smile:

File: apps/backend-functions/src/contract.ts [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        await tx.set(versionSnap.ref, versionDoc);
+      } else {
+        // There is no remaining contract version we delete _meta document and the corresponding publicContract will be deleted
+        console.log(`deleting _meta document for contractId : ${contract.id}`);
+        tx.delete(versionSnap.ref);
+      }
+    }
+
+    /** @dev public contract document is created only it status is OK */
+    if (lastVersionDoc && ['waitingpayment', 'paid', 'accepted'].includes(lastVersionDoc.status)) {
```
Comment (lsenta) [inline apps/backend-functions/src/contract.ts]: define in a const, should be an enum in the firestore model

File: apps/backend-functions/src/contract.ts [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+  return db.runTransaction(async tx => {
+    await tx.delete(db.doc(`publicContracts/${contract.id}`));
+  });
+}
+
+export async function onContractVersionDelete(
+  snap: FirebaseFirestore.DocumentSnapshot,
+): Promise<any> {
+  const contractId = snap.ref.parent.parent ? snap.ref.parent.parent.id : undefined;
```
Comment (lsenta) [inline apps/backend-functions/src/contract.ts]: use `const {contractId} = context.params` it matches params in your firestore ref string  (`/contract/{contractId}/etc`)

File: apps/backend-functions/src/contract.ts [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+  // We retreive the version ID beeing currently removed
```
Comment (lsenta) [inline apps/backend-functions/src/contract.ts]: retrieve, being

File: apps/backend-functions/src/main.ts [hunk 5]
```diff
@@ -153,6 +161,59 @@ export const onMovieDeleteEvent = onDocumentDelete(
...
+//------------------------------------------------
+//   Contracts & Contracts Version Management   //
+//------------------------------------------------
+
+/**
+ * Trigger: when a contract is created
+ */
+export const onContractCreateEvent = onDocumentCreate(
+  'contracts/{contractId}',
+  onContractCreate
```
Comment (lsenta) [inline apps/backend-functions/src/main.ts]: use a single `onDocumentWrite` if you do crete, update and delete

File: apps/backend-functions/src/main.ts [hunk 6]
```diff
@@ -153,6 +161,59 @@ export const onMovieDeleteEvent = onDocumentDelete(
...
+  'contracts/{contractId}',
+  onContractDelete
+)
+
+/**
+ * Trigger: when a contractVersion is created
+ */
+export const onContractVersionCreateEvent = onDocumentCreate(
+  'contracts/{contractId}/versions/{versionId}',
+  onContractVersionCreate
```
Comment (lsenta) [inline apps/backend-functions/src/main.ts]: Same, 3 functions => 1 write

File: apps/backend-functions/src/contract.ts [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+      }
+    }
+
+    /** @dev public contract document is created only it status is OK */
+    if (lastVersionDoc && ['waitingpayment', 'paid', 'accepted'].includes(lastVersionDoc.status)) {
+      const publicContract: PublicContractDocument = {
+        id: contract.id,
+        type: contract.type,
+        titleIds: contract.titleIds,
+        scope: lastVersionDoc ? lastVersionDoc.scope : {},
```
Comment (lsenta) [inline apps/backend-functions/src/contract.ts]: empty scope, doesn't seems to be handled by the type

File: apps/backend-functions/src/contract.ts [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+  return db.runTransaction(async tx => {
+    const publicContractSnap = await tx.get(db.doc(`publicContracts/${contract.id}`));
+
+    // Fetching the current version to compare it against versionToSkip value
+    const versionSnap = await tx.get(db.doc(`contracts/${contract.id}/versions/_meta`));
+    const versionDoc = versionSnap.data();
+    let lastVersionDoc;
+    if (!!versionDoc) {
+      // If the current version is the one to skip/delete (versionToSkip), we update _meta version with versionToSkip - 1
+      const versionToFetch = !versionToSkip || parseInt(versionToSkip, 10) !== parseInt(versionDoc.count, 10) ? versionDoc.count : parseInt(versionToSkip, 10) - 1;
```
Comment (lsenta) [inline apps/backend-functions/src/contract.ts]: too complex, split.

Comment (lsenta) [reply apps/backend-functions/src/contract.ts]: you are relying on the fact that:
- firestore events are triggered in order (delete(x) happens BEFORE delete(x-1)) which is probably not true,
   - you'll might processed delete(x-1) => version = x - 1; then delete(x) => version = x -1 which was deleted right before.
- contract versions are removed in order

Comment (lsenta) [reply apps/backend-functions/src/contract.ts]: Why do you need to store a count to get the last version?
just write the version number in the doc and query the last one. You could also simply store a created time and use that to get the last version.

Counting and assuming ids are sequential and ordered has so many requirements, you're going to hide bug everywhere in the app (back and front). It's costly and unreliable if not done correctly.

Comment (phpgeekfr) [reply apps/backend-functions/src/contract.ts]: Added comments with #1887. We will stay like that until Berlin but we will work on a better solution after.

File: apps/backend-functions/src/contract.ts [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+export async function onContractWrite(
+  change: functions.Change<FirebaseFirestore.DocumentSnapshot>,
+): Promise<any> {
+
+  const before = change.before.data();
+  const after = change.after.data();
+
+  if (!after && !!before) { // Deletion
+    return db.runTransaction(async tx => {
+      await tx.delete(db.doc(`publicContracts/${before.id}`));
```
Comment (lsenta) [inline apps/backend-functions/src/contract.ts]: `db.doc(`publicContracts/${before.id}`).delete()` a transaction is necessary when you have write operations that depends on reads. There's only one operation here.

File: apps/backend-functions/src/contract.ts [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+  const before = change.before.data();
+  const after = change.after.data();
+
+  const contractRef = db.doc(`contracts/${contractId}`);
+  const contractSnap = await contractRef.get();
+  const contract = contractSnap.data() as ContractDocument;
+
+  if (!after && !!before) { // Deletion
+    console.log(`deleting ContractVersion : "${versionId}" for : ${contractId}`);
+    return await transformContractToPublic(contract, versionId);
```
Comment (lsenta) [inline apps/backend-functions/src/contract.ts]: you could just return, `await` is necessary only if you need to capture an error in the current scope (you don't have try catch here).

File: apps/backend-functions/src/contract.ts (hunk 9 above)
Comment (phpgeekfr) [reply apps/backend-functions/src/contract.ts]: Yes you are right thanks

**Pull request merged**